### DISCUSSION TITLE
Confusion with fief and standalone postgresql deployment

### DISCUSSION THREAD

**Discussion by matteius** [fief-dev/fief#162] [Q&A] [answered]
Nice project -- I am starting to get a sense of things, or at least I was last night locally, but now in production I am confused.   I've deployed the docker image and tweaked the environment variables to use my postgresql instance in the cloud.   Things seem to be working, kind-of.  https://members.opensensor.io/admin

First thing I note, there are 5 database tables that get created:
![image](https://user-images.githubusercontent.com/479892/228717982-1f30dce9-ca52-43a3-bb38-6a97661b2b0d.png)

There is one workspace and one alembic db record, the rest is empty.

Ok, so the first question is why isn't it using my env vars for the `FIEF_MAIN_USER_EMAIL` and `FIEF_MAIN_USER_PASSWORD` that I have verified are set on the pods?

Ok no bother, at this point I can register a new account -- but then I am prompted to setup a workspace.  This confuses me because it already has a workspace, but I proceed to set one up which creates weird redirect problems, so I check in the DB and there are two records now.  I tweak the in various ways, delete one.  I get it so I can login -- this is odd.   Also should there be more than 5 database tables?  I feel like there are settings and things I can configure in Fief that I can't find in the database.

Other problems are after I tweak the database records I am unable to register additional users.  So now I drop the tables, restart the deployment and it recreates the tables and I start over.   I haven't figured out how to get past this part -- I think it was impacting the oauth providers I was trying to setup originally but in further debugging I end up breaking the registration flow by registering my user and adding the workspace, one way or another.

**Comment by fief-bailiff** [comment]
Hail, @matteius 👋 Welcome to Fief's kingdom!

Our team will get back to you very soon to help.

In the meantime, take a minute to star our repository ⭐️

![star-fief](https://user-images.githubusercontent.com/1144727/192523552-dc2c31fa-ae94-42fa-9ab4-7d0ba51bb7b0.png)

Farewell!

**Comment by matteius** [comment]
Ah ok I figured out one thing, I had a typo in my env var or `FIEF_MAIN_USER_EMAIL`  -- I had `FIEF_MAIN_USER_EMAILL`  -- so I just dropped tables and restarted and now I have a default user.  Now it seems to not prompt me to create a second workspace when I login 🎉   I'm going to try setting up the external oauth connection again.

Update:  Actually before I set anything else up, I noticed that incognito the registration page is not working now that it created my base user:
![image](https://user-images.githubusercontent.com/479892/228721593-5e2ac63a-e5fc-4b30-818d-752c6af8849a.png)

**Reply by matteius** [reply]
If I navigate to /admin first then there is a link to /register and the register form renders -- if I navigate there directly in a fresh session I get the above image.

**Comment by frankie567** [comment, marked as answer]
Regarding the database, an important thing to understand is that there is a "root" one that manages workspaces and one for each workspace. In the case of PostgreSQL, we leverage schemas for this. So it looks like this:

```mermaid
flowchart LR
    DB[Root]
    WS1[Workspace 1]
    WS2[Workspace 2]
    WSN[Workspace n]
    DB --> WS1
    DB --> WS2
    DB --> WSN
```

Thus, when you start Fief server, it'll create the schema for the "root" database (in `public` schema) and the "main workspace` database (in a schema named after its workspace ID).

If everything goes well, if you go to `/admin/`, you'll be taken to the login page. If you input the credentials set with `FIEF_MAIN_USER_EMAIL` and `FIEF_MAIN_USER_PASSWORD`, you'll have access to your admin dashboard.

---

When you access `/register` directly without any context, it's expected to have the *Invalid login session* error. Actually, the same will happen if you access `/login` directly.

In a platform like Fief, authentication *must* happen within the context of an application.

That's why you need to go through the [OAuth2 flow](https://docs.fief.dev/getting-started/oauth2/) when integrating your application. Under the hood, Fief will create a login session, so it knows where to take the user back when the user has authenticated (or registered).

The admin dashboard uses this flow correctly, that's why it works when you go first through `/admin/`.

**Reply by matteius** [reply]
Thanks -- this was super helpful -- I had to expose my DataGrip to show me the other schemas, now this makes more sense.  I got my site integrated with my Fief server tonight and it works!!!   Only issue I haven't figured out yet with the standard flows is when I login its redirecting back to `https://opensensor.io/[object%20Object]` -- If I remove the `[object%20Object]` it shows I am in fact logged in and can logout.  This is super cool -- nice work on this project!

**Reply by frankie567** [reply]
Great to hear!

> when I login its redirecting back to https://opensensor.io/[object%20Object]

That looks like a JavaScript bug. It seems to be an issue with Next.js 13, I'll look into this!

**Reply by frankie567** [reply]
Fixed in [0.12.2](https://github.com/fief-dev/fief-js/releases/tag/v0.12.2) 👍

**Reply by matteius** [reply]
@frankie567 Can confirm, that fixed it!
